## Races Table:

Importing needed pacakges and making the connection:

In [2]:
from sqlalchemy import create_engine
import pandas as pd

#First, creating an engine and then importing the races file.
#engine = create_engine('postgresql://patrickbovard:localhost@localhost:5432/ponies')

#races_table = pd.read_csv('./Data/races.csv')

#races_table.to_sql('races_table', engine, index=False)

General exploration of races table:

In [3]:
query = '''
SELECT * 
FROM races_table
LIMIT 5
;
'''
df = pd.read_sql(query, engine)

df.head()

race_id        date venue  race_no config  surface  distance         going  \
0        0  1997-06-02    ST        1      A        0      1400  GOOD TO FIRM   
1        1  1997-06-02    ST        2      A        0      1200  GOOD TO FIRM   
2        2  1997-06-02    ST        3      A        0      1400  GOOD TO FIRM   
3        3  1997-06-02    ST        4      A        0      1200  GOOD TO FIRM   
4        4  1997-06-02    ST        5      A        0      1600  GOOD TO FIRM   

  horse_ratings      prize  ...  place_combination3  place_combination4  \
0         40-15   485000.0  ...                 6.0                None   
1         40-15   485000.0  ...                 4.0                None   
2         60-40   625000.0  ...                13.0                None   
3        120-95  1750000.0  ...                10.0                None   
4         60-40   625000.0  ...                 1.0                None   

   place_dividend1  place_dividend2  place_dividend3 place_dividend4  \
0             36.5             25.5             18.0            None   
1             12.5             47.0             33.5            None   
2             23.0             23.0             59.5            None   
3             14.0             24.5             16.0            None   
4             15.5             28.0             17.5            None   

  win_combination1 win_dividend1  win_combination2  win_dividend2  
0                8         121.0              None           None  
1                5          23.5              None           None  
2               11          70.0              None           None  
3                5          52.0              None           None  
4                2          36.5              None           None  

[5 rows x 37 columns]

### Races by Distance:

In [4]:
query = '''
SELECT distance, count(race_id) as number_of_races
FROM races_table
GROUP BY distance
ORDER BY distance ASC
;
'''
df = pd.read_sql(query, engine)

df

distance  number_of_races
0      1000              698
1      1200             2017
2      1400             1081
3      1600              637
4      1650             1095
5      1800              589
6      2000              117
7      2200               90
8      2400               25

### Races by Rating Allowed:

In [5]:
query = '''
SELECT horse_ratings, count(race_id) as number_of_races
FROM races_table
GROUP BY horse_ratings
ORDER BY number_of_races DESC
;
'''
df = pd.read_sql(query, engine)

df

horse_ratings  number_of_races
0          60-40             2196
1          80-60             1786
2          40-15              409
3         100-80              407
4              G              312
5           40-0              271
6          85-60              193
7          40-10              189
8         105-80              154
9          65-40               62
10         95-75               56
11        110-85               46
12        115-90               38
13         60-35               37
14           90+               31
15           95+               24
16         40-20               23
17        110-90               21
18         90-70               21
19         80-55               13
20        115-95               11
21          100+               10
22        100-75               10
23        120-95                9
24           80+                5
25       120-100                4
26         75-55                4
27           85+                3
28        105-85                2
29        110-80                1
30         95-70                1

Based on [this article on theroar.com](https://www.theroar.com.au/2015/07/10/your-complete-guide-to-hong-kong-horse-racing-what-the-fuss-is-all-about/#:~:text=Hong%20Kong%20racing%20is%20predominantly,set%20weights%20and%20penalties%20based.), the 'G' races are called Griffin racing, which includes any horses that haven't raced in Hong Kong before.

### Races by Track Condition:

In [6]:
query = '''
SELECT going, count(race_id) as number_of_races
FROM races_table
GROUP BY going
ORDER BY going ASC
;
'''
df = pd.read_sql(query, engine)

df

going  number_of_races
0              FAST               70
1              GOOD             4123
2      GOOD TO FIRM             1613
3  GOOD TO YIELDING              333
4              SLOW               15
5              SOFT               14
6          WET FAST               20
7          WET SLOW               61
8          YIELDING               86
9  YIELDING TO SOFT               14

### Races by Venue and Configuration:

In [7]:
query = '''
SELECT venue, config, count(race_id) as number_of_races_in_config
FROM races_table
GROUP BY venue, config
ORDER BY venue ASC
;
'''
df = pd.read_sql(query, engine)

df

venue config  number_of_races_in_config
0    HV      B                        658
1    HV      C                        614
2    HV    C+3                        343
3    HV      A                        728
4    ST    A+3                        563
5    ST      A                       1481
6    ST      B                        161
7    ST    B+2                        600
8    ST      C                        560
9    ST    C+3                        641

Per the dataset, HV is the Happy Valley Racetrack, ST is Shatin.  Looks like 

## Runs Table:

In [8]:
#Now, the runs file:
#runs_table = pd.read_csv('./Data/runs.csv')

#runs_table.to_sql('runs_table', engine, index=False)

Going to explore the runs table a bit, see how many horses there are, data by age, country, etc.

In [51]:
query = '''
SELECT * 
FROM runs_table
LIMIT 20
;
'''
df = pd.read_sql(query, engine)

df.head(20)

race_id  horse_no  horse_id  result  won  lengths_behind  horse_age  \
0         0         1      3917      10  0.0            8.00          3   
1         0         2      2157       8  0.0            5.75          3   
2         0         3       858       7  0.0            4.75          3   
3         0         4      1853       9  0.0            6.25          3   
4         0         5      2796       6  0.0            3.75          3   
5         0         6      3296       3  0.0            1.25          3   
6         0         7       911      12  0.0            9.50          3   
7         0         8      2170       1  1.0            0.00          3   
8         0         9      1730      13  0.0            9.75          3   
9         0        10      2998      14  0.0          999.00          3   
10        0        11      1733       2  0.0            0.75          3   
11        0        12      2617       4  0.0            2.25          3   
12        0        13       727      11  0.0            8.25          3   
13        0        14       306       5  0.0            2.75          3   
14        1         1      2397      12  0.0           10.00          3   
15        1         2      4181      10  0.0            6.75          3   
16        1         3      2814       8  0.0            3.75          3   
17        1         4       905       3  0.0            2.25          3   
18        1         5      3915       1  1.0            0.00          3   
19        1         6      3865      11  0.0            8.50          3   

   horse_country horse_type  horse_rating  ...  time2  time3  time4  time5  \
0            AUS    Gelding            60  ...  21.59  23.86  24.62   None   
1             NZ    Gelding            60  ...  21.99  23.30  23.70   None   
2             NZ    Gelding            60  ...  21.59  23.90  24.22   None   
3            SAF    Gelding            60  ...  21.83  23.70  24.00   None   
4             GB    Gelding            60  ...  21.75  23.22  23.50   None   
5             NZ    Gelding            60  ...  22.03  22.90  23.57   None   
6             NZ    Gelding            60  ...  21.59  23.94  25.09   None   
7            AUS    Gelding            60  ...  21.87  23.58  23.06   None   
8             NZ    Gelding            60  ...  21.71  23.90  24.94   None   
9            AUS       Mare            60  ...  22.31  24.38  30.46   None   
10            NZ    Gelding            60  ...  21.75  23.54  23.51   None   
11           AUS    Gelding            60  ...  21.91  23.30  23.52   None   
12           USA    Gelding            60  ...  21.79  23.82  24.52   None   
13           AUS    Gelding            60  ...  21.75  23.90  23.66   None   
14            NZ    Gelding            60  ...  22.92  24.05    NaN   None   
15            NZ    Gelding            60  ...  22.84  23.58    NaN   None   
16           AUS      Horse            60  ...  22.92  23.29    NaN   None   
17           AUS    Gelding            60  ...  22.84  22.75    NaN   None   
18           AUS    Gelding            60  ...  22.80  23.06    NaN   None   
19            NZ    Gelding            60  ...  23.12  24.32    NaN   None   

    time6  finish_time  win_odds  place_odds trainer_id jockey_id  
0    None        83.92       9.7         3.7        118         2  
1    None        83.56      16.0         4.9        164        57  
2    None        83.40       3.5         1.5        137        18  
3    None        83.62      39.0        11.0         80        59  
4    None        83.24      50.0        14.0          9       154  
5    None        82.83       7.0         1.8         54        34  
6    None        84.15      99.0        28.0         55       149  
7    None        82.64      12.0         3.6         47       183  
8    None        84.20      38.0        13.0         75       131  
9    None        92.20      39.0        12.0        109       145  
10   None        82.77       8.6         

### Number of horses:

In [10]:
query = '''
SELECT horse_id, count(result) AS races_run
FROM runs_table
GROUP BY horse_id
ORDER BY races_run DESC
;
'''
df = pd.read_sql(query, engine)

df

horse_id  races_run
0         3887         92
1          493         88
2         2382         83
3           70         77
4          779         77
...        ...        ...
4400      3640          1
4401       478          1
4402      2460          1
4403      2881          1
4404      1130          1

[4405 rows x 2 columns]

4405 unique horses than ran at least one race.

### Horses by age:

In [11]:
query = '''
SELECT horse_age, count(horse_id) AS horses_in_age, avg(result) as Avg_result, avg(draw) as Avg_draw
FROM runs_table
GROUP BY horse_age
ORDER BY horse_age ASC
;
'''
df = pd.read_sql(query, engine)

df

horse_age  horses_in_age  avg_result  avg_draw
0          2           2098    7.249285  7.060534
1          3          61212    6.926322  6.885986
2          4           8007    6.143499  6.850506
3          5           4775    6.480209  6.776335
4          6           2244    7.127451  6.784759
5          7            839    7.117998  6.724672
6          8            225    7.666667  6.968889
7          9             41    6.804878  6.707317
8         10              6    3.166667  6.833333

This makes sense, since most thoroughbred horseracing is on horses ages 2-3.  Overall there is a largae range of ages, but the vast majority are 2-4. 

### Horses by Country:

In [12]:
query = '''
SELECT horse_country, count(horse_id) AS horses_from_country, avg(result) as Avg_result, avg(draw) as Avg_draw
FROM runs_table
GROUP BY horse_country
ORDER BY horse_country ASC
;
'''
df = pd.read_sql(query, engine)

df

horse_country  horses_from_country  avg_result  avg_draw
0            ARG                  127    8.511811  7.204724
1            AUS                29906    6.764830  6.893132
2            BRZ                   18    8.722222  6.944444
3            CAN                   91    8.571429  6.901099
4             FR                 1257    7.379475  7.064439
5             GB                 6329    6.873914  6.732817
6            GER                  357    7.204482  6.963585
7             GR                   33    7.848485  5.363636
8            IRE                10473    6.902607  6.770839
9            ITY                   66    7.696970  7.075758
10           JPN                   93    7.322581  6.440860
11            NZ                27497    6.791032  6.938466
12           SAF                  685    7.103650  6.671533
13           SPA                    2   13.000000  7.000000
14           USA                 2499    7.259704  6.742697
15           ZIM                   12    8.666667  8.166667
16          None                    2    8.000000  6.500000

### Different Gear Types by the Horse:

What is the gear a horse may have?

In [21]:
query = '''
SELECT horse_gear, count(horse_id) AS horses_with_gear
FROM runs_table
GROUP BY horse_gear
ORDER BY horse_gear ASC
;
'''
df = pd.read_sql(query, engine)

df

horse_gear  horses_with_gear
0           --             57373
1            B              3579
2           B-               186
3       B-/BO1                 2
4       B-/CP1                12
..         ...               ...
817      XB/V1                 1
818        XB1                46
819     XB1/B2                 2
820     XB1/H2                 1
821        XB2                 1

[822 rows x 2 columns]

May need to look into this more, but it seems a rough guide is [here on the HKJC site](https://racing.hkjc.com/racing/english/racing-info/reg_approved_gear.aspx).  

Includes:  
- Blinkers (B, BO)  
- Visor (V)  
- Pacifiers (P, PC, PS)
- Cheekpieces (CP, CO), Shadow Roll (SR), Browband (SB)
- Hood (H), Ear Plugs (E)
- Etc.

### Horses by Type (Sex):

In [13]:
query = '''
SELECT horse_type, count(horse_id) AS horses_in_type, avg(result) as Avg_result, avg(draw) as Avg_draw
FROM runs_table
GROUP BY horse_type
ORDER BY horses_in_type DESC
;
'''
df = pd.read_sql(query, engine)

df

horse_type  horses_in_type  avg_result  avg_draw
0    Gelding           75503    6.836841  6.884058
1      Brown            2044    6.825342  6.793053
2      Horse            1093    6.704483  6.453797
3       Colt             305    7.337705  6.809836
4       Mare             238    7.096639  6.722689
5        Rig             154    6.935065  7.370130
6       Roan              48    6.708333  6.854167
7      Filly              43    8.279070  6.976744
8       Grey              17    8.000000  6.941176
9       None               2    8.000000  6.500000

- Looks like there's a mix of different types:  
    - Sex:
        - Gelding - "fixed" male horse
        - Mare - adult female horse
        - Filly - young female horse
        - Horse - adult male horse
        - Colt - young male horse
        - Rig - male horse that "appears" like a gelding but isn't
    - Color:  
        - Brown
        - Roan
        - Grey
    - Other:
        - None  

May need to consider only focusing on Geldings, since some attributes like weight, etc. would dfeinitely be influenced by sex.

### Jockers and Trainers

I only have id's for jockeys and trainers, and no other info.  I'd like to see how many of each there are, as possible features correlated with success.  Is there a Hong Kong [Bob Baffert](https://en.wikipedia.org/wiki/Bob_Baffert)?

Trainers - how many are there, and who is in the top 10?

In [22]:
#Trainers - Numbers
query = '''
SELECT trainer_id, COUNT(*) as times_used
FROM runs_table
GROUP BY trainer_id
ORDER BY times_used DESC
;
'''
df = pd.read_sql(query, engine)

df

trainer_id  times_used
0           137        4739
1            97        4600
2            75        4578
3            47        4482
4             7        4109
..          ...         ...
171         145           1
172         129           1
173         148           1
174          58           1
175          87           1

[176 rows x 2 columns]

In [33]:
#Trainers - Success (w/ a few races)
query = '''
SELECT trainer_id, avg(result) as Avg_Result, COUNT(*) as Number_of_races
FROM runs_table
GROUP BY trainer_id
HAVING COUNT(*) > 20
ORDER BY Avg_Result ASC
LIMIT 10
;
'''
df = pd.read_sql(query, engine)

df

trainer_id  avg_result  number_of_races
0         118    5.604069             3834
1           7    5.780725             4109
2          97    6.327174             4600
3         137    6.483857             4739
4          80    6.540575             3167
5          27    6.684291             1967
6          60    6.719735             2569
7          29    6.722900             2786
8          63    6.806923             1300
9         164    6.807519             3777

Jockeys - how many are there, and who are the top 10?

In [24]:
#Jockeys - Numbers
query = '''
SELECT jockey_id, COUNT(*) as times_used
FROM runs_table
GROUP BY jockey_id
ORDER BY times_used DESC
;
'''
df = pd.read_sql(query, engine)

df

jockey_id  times_used
0            2        4719
1           64        4381
2           18        3985
3           63        3767
4           34        3590
..         ...         ...
181        170           1
182         84           1
183        178           1
184        148           1
185        116           1

[186 rows x 2 columns]

In [34]:
#Jockeys - Success (w/ a few races)
query = '''
SELECT jockey_id, avg(result) as Avg_Result, COUNT(*) as Number_of_races
FROM runs_table
GROUP BY jockey_id
HAVING COUNT(*) > 20
ORDER BY Avg_Result ASC
LIMIT 10
;
'''
df = pd.read_sql(query, engine)

df

jockey_id  avg_result  number_of_races
0         95    4.291139             1975
1          2    5.247298             4719
2        115    5.386822             1290
3        134    5.537975              158
4         18    5.768130             3985
5         64    5.784296             4381
6         96    5.822917              192
7         11    5.966292              445
8          9    6.049327              223
9         82    6.219512              205

Unfortunately the data set has no way to tie these to actual people (names, etc.), but it is good to know.

# Basic Joins:

Next step, I want to do some basic table joins to combine the data from the specific race onto the run:

In [42]:
query = '''
SELECT ru.horse_id, ru.horse_type, ru.horse_age, ru.result, ru.draw, ru.declared_weight, ru.actual_weight, ru.horse_rating, ra.distance, ra.venue, ra.config, ra.horse_ratings
FROM runs_table as ru
LEFT JOIN races_table as ra
    ON ru.race_id = ra.race_id
;
'''
df = pd.read_sql(query, engine)

df.sample(20)

horse_id horse_type  horse_age  result  draw  declared_weight  \
59254      2954    Gelding          3       3     5           1023.0   
29107       532    Gelding          3       6    11           1119.0   
76976      1908    Gelding          6       3     1           1111.0   
78802      1791    Gelding          4       1    10           1139.0   
49467      1949    Gelding          5      11     7           1102.0   
53051      3328    Gelding          3       2     7           1204.0   
65944        70    Gelding          8       7     1           1088.0   
3294       3283      Brown          3       6    14           1076.0   
27005       830    Gelding          3       3     1            990.0   
15445       830    Gelding          3       2     6           1003.0   
62840      1947    Gelding          7      11     5           1192.0   
55141      1722    Gelding          5       7     3            989.0   
6593       2488    Gelding          3      12     7           1065.0   
76764      1801    Gelding          4       8     4           1086.0   
53499      1363    Gelding          3      13     6           1050.0   
10064      3424    Gelding          3       1    10           1102.0   
23911      3835    Gelding          3       7     1           1073.0   
24002      2635    Gelding          3       3    10           1052.0   
48869      1365    Gelding          3       4     9           1087.0   
24661       336    Gelding          3       2     1           1176.0   

       actual_weight  horse_rating  distance venue config horse_ratings  
59254            129            60      1200    HV      C         80-60  
29107            126            60      1650    HV      A        100-80  
76976            115            29      1200    HV    C+3         40-15  
78802            126            72      1650    HV    C+3         80-60  
49467            116            81      1200    ST    B+2        100-80  
53051            130            60      1400    ST      C         90-70  
65944            113            64      1650    HV    C+3         80-60  
3294             111            60      1600    ST    A+3         80-60  
27005            127            60      1800    HV      A         80-60  
15445            129            60      1800    ST    C+3         80-60  
62840            111            20      1800    ST      A          40-0  
55141            119            66      1800    ST      A         80-60  
6593             111            60      1200    ST      A         60-40  
76764            122            51      1200    HV      C         60-40  
53499            118            60      1600    ST    C+3         80-60  
10064            119            60      1600    ST      A         95-75  
23911            108            60      2200    HV    C+3         80-60  
24002            120            60      1200    ST    B+2         80-60  
48869            133            60      1400    ST    C+3         40-15  
24661            127            60      1200    ST      B         60-40

Actual weight is the weight the horse is carrying (i.e. to handicap the race), declared weight is the weight of the horse+jockey.  

Checking out one race to see how it's distributed:

In [58]:
query = '''
SELECT ru.horse_id, ru.horse_type, ru.horse_age, ru.result, ru.draw, ru.declared_weight, ru.actual_weight, ru.actual_weight + ru.declared_weight AS combined_weight, ru.horse_rating, ra.distance, ra.venue, ra.config, ra.horse_ratings
FROM runs_table as ru
LEFT JOIN races_table as ra
    ON ru.race_id = ra.race_id
WHERE ra.race_id = 9
ORDER BY ru.draw ASC
;
'''
df = pd.read_sql(query, engine)

df

horse_id horse_type  horse_age  result  draw  declared_weight  \
0       2053    Gelding          3      10     1           1079.0   
1        908    Gelding          3      12     2           1148.0   
2       1444    Gelding          3       8     3           1210.0   
3       4043    Gelding          3      14     4           1054.0   
4       3306    Gelding          3      13     5           1130.0   
5        541    Gelding          3       7     6           1074.0   
6       3683    Gelding          3       9     7           1186.0   
7       3422    Gelding          3       4     8           1040.0   
8       3842    Gelding          3       6     9           1087.0   
9       3451    Gelding          3       1    10           1046.0   
10       477    Gelding          3       5    11           1071.0   
11       540    Gelding          3      11    12           1050.0   
12      3117      Brown          3       2    13           1146.0   
13      3668    Gelding          3       3    14           1205.0   

    actual_weight  combined_weight  horse_rating  distance venue config  \
0             119           1198.0            60      1200    ST      A   
1             120           1268.0            60      1200    ST      A   
2             129           1339.0            60      1200    ST      A   
3             116           1170.0            60      1200    ST      A   
4             133           1263.0            60      1200    ST      A   
5             131           1205.0            60      1200    ST      A   
6             120           1306.0            60      1200    ST      A   
7             124           1164.0            60      1200    ST      A   
8             115           1202.0            60      1200    ST      A   
9             133           1179.0            60      1200    ST      A   
10            113           1184.0            60      1200    ST      A   
11            126           1176.0            60      1200    ST      A   
12            111           1257.0            60      1200    ST      A   
13            124           1329.0            60      1200    ST      A   

   horse_ratings  
0          80-60  
1          80-60  
2          80-60  
3          80-60  
4          80-60  
5          80-60  
6          80-60  
7          80-60  
8          80-60  
9          80-60  
10         80-60  
11         80-60  
12         80-60  
13         80-60

In [37]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 79447 entries, 0 to 79446
Data columns (total 12 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   horse_id         79447 non-null  int64  
 1   horse_type       79445 non-null  object 
 2   horse_age        79447 non-null  int64  
 3   result           79447 non-null  int64  
 4   draw             79447 non-null  int64  
 5   declared_weight  79447 non-null  float64
 6   actual_weight    79447 non-null  int64  
 7   horse_rating     79447 non-null  int64  
 8   distance         79447 non-null  int64  
 9   venue            79447 non-null  object 
 10  config           79447 non-null  object 
 11  horse_ratings    79447 non-null  object 
dtypes: float64(1), int64(7), object(4)
memory usage: 7.3+ MB


Let's see if certain horse types tend to do better at either track, Happy Valley or Sha Tin:

In [19]:
query = '''
SELECT ra.venue, ru.horse_type, avg(ru.horse_age) as Avg_age, avg(ru.result) as Avg_Finish, count(ru.horse_id) as Number_of_Runs
FROM runs_table as ru
LEFT JOIN races_table as ra
    ON ru.race_id = ra.race_id
GROUP BY ra.venue, ru.horse_type
ORDER BY ra.venue ASC, Avg_Finish ASC
;
'''
df = pd.read_sql(query, engine)

df

venue horse_type   avg_age  avg_finish  number_of_runs
0     HV        Rig  3.175000    5.950000              40
1     HV       Mare  3.372549    6.078431              51
2     HV    Gelding  3.398420    6.384174           26324
3     HV      Horse  3.279863    6.392491             293
4     HV      Brown  3.092541    6.520718             724
5     HV       Colt  3.000000    7.325000              40
6     HV       Grey  3.000000    7.400000              10
7     HV       Roan  4.142857    7.428571               7
8     HV       None  3.000000    8.000000               2
9     HV      Filly  3.000000    8.100000              10
10    ST       Roan  5.048780    6.585366              41
11    ST      Horse  3.178750    6.818750             800
12    ST      Brown  3.115909    6.992424            1320
13    ST    Gelding  3.321804    7.079139           49179
14    ST        Rig  3.245614    7.280702             114
15    ST       Colt  2.958491    7.339623             265
16    ST       Mare  3.358289    7.374332             187
17    ST      Filly  2.969697    8.333333              33
18    ST       Grey  3.000000    8.857143               7

## Close the Session RUN AT END:

In [ ]:
engine.dispose()